### PR TITLE
destinations v2: snowflake: single-threaded T+D per stream

### LINKED ISSUE #30048: Async support for T&D

### PR DESCRIPTION
**Pull request by edgao**
Closes https://github.com/airbytehq/airbyte/issues/30048

Add locks to both T+D and raw table commits for snowflake/bigquery:
* any number of `copyIntoTableFromStage` calls can run concurrently
* `typeAndDedupe` can only run one at a time

And reenables mid-sync T+D execution for both snowflake, and bigquery GCS.

Async folks - PTAL at GeneralStagingFunctions and let me know if it looks problematic for destination-bigquery.

Example logs:
image
* One thread tries to start T+D (the `Waiting for raw table writes to pause` log message at the top)
* In the meantime, another thread is flushing to the raw tables, so nothing happens
* The flush finishes, so the first thread is able to start running T+D (`Attempting typing and deduping` - note the timestamp is a few seconds after the first message)
* The second thread also tries to trigger T+D, but skips it because T+D is already running (`Another thread is already trying to run typing and deduping`)

### CODE REVIEW

File: airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java [hunk 1]
```diff
@@ -127,16 +133,18 @@ public void prepareTables() throws Exception {
   public void typeAndDedupe(final String originalNamespace, final String originalName) throws Exception {
     LOGGER.info("Attempting typing and deduping for {}.{}", originalNamespace, originalName);
     final var streamConfig = parsedCatalog.getStream(originalNamespace, originalName);
-    if (streamsWithSuccesfulSetup.stream()
-        .noneMatch(streamId -> streamId.originalNamespace().equals(originalNamespace) && streamId.originalName().equals(originalName))) {
-      // For example, if T+D setup fails, but the consumer tries to run T+D on all streams during close,
-      // we should skip it.
-      LOGGER.warn("Skipping typing and deduping for {}.{} because we could not set up the tables for this stream.", originalNamespace, originalName);
-      return;
+    synchronized(tdLocks.get(streamConfig.id())) {
```
[the remark this answers is not in the document]
Comment (edgao) [reply airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java:136]: good point. synchronized means they'll all wait their turn. probably need to use a stronger concurrency object here.

we'll still need a way to force T+D to run at the end of a sync (i.e. using the wait-my-turn behavior). There's probably something fancy to do there, e.g. track the last time we started a T+D run vs the last time we committed raw data and skip if there's no new data... but that can be a future improvement

[the remark this answers is not in the document]
Comment (evantahler) [reply airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java:136]: > we'll still need a way to force T+D to run at the end of a sync
for some reason I thought we hooked into the shutdown behavior for that 🤷

[the remark this answers is not in the document]
Comment (edgao) [reply airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java:136]: > shutdown behavior

we do. I think I was just mistaken about this - was worried about a case where we have an in-flight T+D run while the shutdown hook is running, which I don't think is possible.

[the remark this answers is not in the document]
Comment (edgao) [reply airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java]: added an explicit `mustRun` param anyway, because I was having a hard time writing a comment explaining why it isn't necessary :shrug: and if anyone ever gets clever and runs T+D in a separate thread, then we probably need it anyway

**Comment by edgao** [comment]
turning off draft mode to trigger CI, please don't review yet

File: airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/GeneralStagingFunctions.java [hunk 2]
```diff
@@ -77,8 +79,23 @@ public static void copyIntoTableFromStage(final JdbcDatabase database,
...
+        stagingOperations.copyIntoTableFromStage(database, stageName, stagingPath, stagedFiles,
+            tableName, schemaName);
+      } finally {
+        rawTableInsertLock.unlock();
+      }
+
+      final AirbyteStreamNameNamespacePair streamId = new AirbyteStreamNameNamespacePair(streamName, streamNamespace);
+      if (!typerDeduperValve.containsKey(streamId)) {
+        typerDeduperValve.addStream(streamId);
+      }
```
Comment (tryangul) [inline airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/GeneralStagingFunctions.java:94]: Maybe `putIfAbsent` is better here to avoid races

File: airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/GeneralStagingFunctions.java [hunk 3]
```diff
@@ -77,8 +79,23 @@ public static void copyIntoTableFromStage(final JdbcDatabase database,
...
+      if (typerDeduperValve.readyToTypeAndDedupe(streamId)) {
+        typerDeduper.typeAndDedupe(streamId.getNamespace(), streamId.getName(), false);
```
Comment (tryangul) [inline airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/GeneralStagingFunctions.java:96]: Is there any issue if it is run concurrently for the same stream from multiple threads?

Comment (edgao) [reply airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/GeneralStagingFunctions.java:96]: typeAndDedupe handles that internally, as long as callers use `getRawTableInsertLock().lock()` correctly (https://github.com/airbytehq/airbyte/pull/29878/files#diff-7e472d19a2456a0de3d5794f56cfe9c53ed7ba1af8001bb8cacebdde488e8063R147)

File: airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/GeneralStagingFunctions.java (hunk 2 above)
Comment (edgao) [reply airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/GeneralStagingFunctions.java:94]: oh that's a good call, though I think the valve is very nonthreadsafe anyway

Comment (edgao) [reply airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/GeneralStagingFunctions.java]: done in https://github.com/airbytehq/airbyte/pull/29878/commits/9270ab670d933f84f38eeb0ce51bf411fa1c4e6e

**Comment by edgao** [comment]
prerelease publish running https://github.com/airbytehq/airbyte/actions/runs/6028370343

**Review by jbfbell — COMMENTED** [review]
@edgao do you have a sense of the performance impact for this vs. only T&D'ing at the end?

**Comment by edgao** [comment]
destination-snowflake will slow down, and bigquery will remain equally slow. This PR makes T+D completely block raw table writes, whereas previously you could write new raw records while T+D runs (which is wrong, but faster 😛 ). Async snowflake was making use of that, but bigquery at least was already single-threaded anyway.

I'm inclined to close this PR, unless we want to keep the early-sync T+D run.

**Comment by evantahler** [comment]
> I'm inclined to close this PR, unless we want to keep the early-sync T+D run.
I would like to keep the early T&D run... maybe we do it only the first batch... but I don't want to write off "see your data mid-sync" just yet

**Comment by edgao** [comment]
oh, git massively screwed up that merge 🤦 typeAndDedupeTask needs to have the lock logic, and typeAndDedupe(String, String) needs to call typeAndDedupeTask

File: airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java [hunk 4]
```diff
@@ -180,11 +223,38 @@ public CompletableFuture<Optional<Exception>> typeAndDedupeTask(StreamConfig str
               originalName);
           return Optional.empty();
         }
-        final String suffix = getFinalTableSuffix(streamConfig.id());
-        final String sql = sqlGenerator.updateTable(streamConfig, suffix);
-        destinationHandler.execute(sql);
+
+        final boolean run;
```
Comment (jbfbell) [inline airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java:227]: lines 227-255 are identical to 179-206 (give or take some spacing) would suggest either

- creating a helper method called by both
- having `typeAndDedupeTask` just call `typeAndDedupe` (line 170)

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryStagingConsumerFactory.java [hunk 5]
```diff
@@ -81,15 +83,15 @@ public AirbyteMessageConsumer create(final JsonNode config,
 
   // TODO Commenting this out for now since it slows down syncs
   private CheckedConsumer<AirbyteStreamNameNamespacePair, Exception> incrementalTypingAndDedupingStreamConsumer(final TyperDeduper typerDeduper) {
-    // final TypeAndDedupeOperationValve valve = new TypeAndDedupeOperationValve();
+    final TypeAndDedupeOperationValve valve = new TypeAndDedupeOperationValve();
```
Comment (jbfbell) [inline airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryStagingConsumerFactory.java:86]: chatted about this a bit offline, but generally concerned about the performance impact of enabling this not asynchronously. Will run some tests locally to confirm

Comment (jbfbell) [reply airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryStagingConsumerFactory.java:86]: ran a test locally - source postgres to bq gcs staging w/ 30 streams all full refresh, commenting out lines 91-94 sync time was ~12.5min leaving this as is and running the same sync is ~20 minutes at the very least we should comment out these lines otherwise I can add the commit we talked about @edgao to async this for bq

File: airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java (hunk 4 above)
Comment (edgao) [reply airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java:227]: ugh, I thought I fixed this. this was a bad merge, I'd intended to do your second bullet

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryStagingConsumerFactory.java [hunk 6]
```diff
@@ -81,15 +83,15 @@ public AirbyteMessageConsumer create(final JsonNode config,
...
-      // typerDeduper.typeAndDedupe(streamId.getNamespace(), streamId.getName());
-      // valve.updateTimeAndIncreaseInterval(streamId);
-      // }
+      if (!valve.containsKey(streamId)) {
+        valve.addStream(streamId);
+      }
+      if (valve.readyToTypeAndDedupe(streamId)) {
+        typerDeduper.typeAndDedupe(streamId.getNamespace(), streamId.getName(), false);
+        valve.updateTimeAndIncreaseInterval(streamId);
+      }
```
Comment (jbfbell) [inline airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryStagingConsumerFactory.java:94]:
```suggestion
//      if (valve.readyToTypeAndDedupe(streamId)) {
        // typerDeduper.typeAndDedupe(streamId.getNamespace(), streamId.getName(), false);
//        valve.updateTimeAndIncreaseInterval(streamId);
//      }
```

**Review by jbfbell — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryStagingConsumerFactory.java (hunk 5 above)
Comment (edgao) [reply airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryStagingConsumerFactory.java:86]: +1 on commenting this out again. for now I think async destinations should get the incremental T+D since it's relatively free, but maybe we just skip incremental T+D for non-async destinations

destination-bigquery async rollout is probably happening soon, so we can uncomment it once that happens

File: airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java (hunk 4 above)
Comment (edgao) [reply airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/DefaultTyperDeduper.java:203]: fixed, how's this look?

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryStagingConsumerFactory.java (hunk 5 above)
Comment (edgao) [reply airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryStagingConsumerFactory.java:86]: fixed.

Comment (jbfbell) [reply airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryStagingConsumerFactory.java:86]: re: commenting out those lines
ran another test locally where I uncommented these lines now that `typeAndDedupe` for individual streams is using the type-and-dedupe thread pool, and it has similar performance to running this without using the separate thread pool.
TL;dr stays commented out for now until BigQuery GCS is asyncified because it slows down the overall sync

**Review by jbfbell — APPROVED** [review]
